### DISCUSSION TITLE
Why Mediators (or equivalent) must be part of the Trust Spanning Protocol

### DISCUSSION THREAD

**Discussion by TelegramSam** [trustoverip/trust-spanning-protocol#37] [General]
Frequently, the Mediation feature of DIDComm is brought up as unnecessary extra baggage as it relates to the Trust Spanning Protocol. I think some type of mediation feature MUST be part of the TSP, and I'll explain why.

Most internet protocols in common use assume reasonable levels of high availability. This makes sense in an Internet world fundamentally divided into clients and servers.  Clients are often occasionally connected and servers are mostly highly available.

I cringe to imagine a Trust Spanning Protocol without supporting the direct involvement of People and the devices they use. We have an incredible number of personal computing devices available in today's world, and very few of them are able to communicate as first-class citizens on an Internet where the majority of the protocols relegate them to the role of a client in all interactions.

The only way to allow these network-challenged devices to be first-class citizens is to allow them balanced participation in the protocol(s) we use, and that is with some sort of a message relay. These reasons are why DIDComm included Routing and Mediation as a top level feature in the spec. It's the same reason we'll need something of the sort, even if we don't use DIDComm.

**Comment by mwherman2000** [comment]
I totally agree and have included support for (a) _disconnected agents_ and (b) _automatic intelligent routing_ as a Layer 1 Foundation Services Subprotocol - a specialization of the Layer 2 Trust Spanning Layer Base Protocol.

**Comment by talltree** [comment]
@TelegramSam I'm not a protocol expert at the level of the others in this task force, so I can only say that my intuition is that you are right. What I don't know is whether routing and mediation are:

- functions of ToIP `intermediary systems` (as defined in the [ToIP Technology Architecture Specification V1.0](https://github.com/trustoverip/TechArch/blob/main/spec.md)) that are supported by specific features in the TSP, OR
- a layer on top of (or, as @SmithSamuelM calls it, an `overlay` on) the basic `verifiable messaging` capability of TSP "sublayer 1" as [Proposal #2](https://github.com/trustoverip/trust-spanning-protocol/discussions/19) from @dhh1128 suggested).

I'm hoping that's part of what we will be figuring out in the consolidation stage.

**Reply by TelegramSam** [reply]
In the 4 layer stack, I believe that layer 2 needs to have routing. If you split layer 2 into sublayers, I think it's a higher layer than the basics of messaging. I feel pretty strongly about it not being in layer 3 - those trust tasks just need the ability to communicate with people and their devices without worrying about how that happens..

**Reply by talltree** [reply]
@TelegramSam I agree that routing should be accomplished below the trust task layer, i.e., it needs to be something any trust task protocol can get from Layer 2 (or below).

**Reply by dhh1128** [reply]
>In the 4 layer stack, I believe that layer 2 needs to have routing... I think it's a higher layer than the basics of messaging...

Routing is absolutely essential. However, it is interesting that DIDComm v1 and v2 both model routing as a higher-level protocol. Routing is a trust task, I think, in a certain sense. But it is a trust task that is special in two ways. First, it consumes and produces trust between a sender and an intermediary and a receiver, not just between a sender and a receiver. That property is quite important, and quite unique. Second, the ability to behave in this way is something we want to require all participants in the ecosystem to support. They may not want to use it themselves, but they MUST be capable of talking to parties who use it. If an implementation can claim to support TSP, yet not talk to parties who require routing, that implementation places sometimes-connected devices like mobile phones on unequal footing with servers and undermines a key power balance that needs to exist if we really want to solve trust on the internet.

This need for everyone to support it is the main argument, in my mind, for not putting it in layer 3. But it is a bit hazy to me whether the premise justifies the conclusion. Could we require something without putting it in layer 2? I don't know...

**Comment by mwherman2000** [comment]
*This is going to forever be a circular discussion until we get _absolute clarity_ of what Layer 1 is all about.* @TelegramSam @talltree 

The layers are software layers. They actually have to be ...like in the ISO OSI model: software layers bound or glued together by protocols.

![](https://user-images.githubusercontent.com/6101736/227157498-a101e072-1a7b-44b7-9182-2a9defc1df0e.png)

In Proposal 4 Appendix A version 0.30, I've had to reclarify this myself. Layer 1 is the _Layer 1 Trust Foundation **Services**_ ...a layer of trust foundation software _services_. The _DIDComm Network_, a network of DIDComm-connected DIDComm Mediators, is an example of a _trust foundation software service_. (The _DID Registry_ is a second example. Both of these software services live in Layer 1 Trust Foundation Services.)

The Layer 1 Subprotocols for accessing each of these software services are specializations of the DIDComm Layer 2 Trust Spanning Layer Base Protocol.  

So in the sense of being a specialization or derivation of the Base Protocol, they "are higher level" but these Subprotocols are used to access "lower-level trust foundation software services".  **This is a very important concept to wrap your head around** ...the difference between (a) being a specialization of the Base Protocol and (b) accessing "lower-level trust foundation software services".

![image](https://user-images.githubusercontent.com/6101736/227720058-eed158a6-491d-4082-a15f-9ff974f9b674.png)

**Reply by dhh1128** [reply]
@mwherman2000 : I nod my head a most of what you say, but I want to poke a bit at one sentence:

>The Layer 1 Subprotocols for accessing each of these software services are specializations of the DIDComm Layer 2 Trust Spanning Layer Base Protocol.

You say this as if it is an immutable fact. I think it is not quite that strong. In my mind it is conceivable that foundational stuff in layer 1 is accessed using a layer 2 protocol, but I don't think it is guaranteed that this will be the case in all situations. Concrete example: a ledger may exist at layer 1, and may have a protocol by which parties can interact with it -- and that protocol might or might not be based on DIDs and encryption with the keys bound to DIDs. Either possibility seems viable to me. If the ledger *does* use DIDs and DID keys to encrypt, then we could say that the means of accessing it is layer-2-based; if it *does not*, then I think we would say that the means of accessing it is not layer-2-based -- but either way, the ledger still constitutes a layer 1 construct. Right? (Someone might say that layer 1 can't ever depend on layer 2, else we would have circularity. However, I think this is wrong. We only have circularity if layer 1 depends on the very layer 2 that depends on that particular layer 1 construct. If the DIDs differ, or the trust bases differ, there is no circularity in instance, only circularity in kind...)

**Reply by mwherman2000** [reply]
> Right?

Sort of but mostly not really. 

In the Web 7.0 DIDComm-ARM, any instance of a Layer 1 Trust Foundation Service that doesn't expose a native DIDComm service endpoint (e.g. a DID Registry service that only supports DNS as it's native service endpoint protocol) is automatically assumed to have a DIDComm Agent Gateway generated that sits in front of and wraps the non-DIDComm-enabled service.

Everything is DIDComm enabled. 100.0000%.

Everything...especially in Layer 1...uses and/or exposes itself using a DIDComm protocol (a Subprotocol in the case of Layer 1 Trust Foundation Services).

**Reply by TelegramSam** [reply]
I think layer 1 should be labeled _cryptographic_ trust. I don't think DIDComm lives at layer 1.

**Reply by mwherman2000** [reply]
DIDComm Messaging is the secure, private protocol that binds the layers together...all of them. It's the glue that binds the layers together and as such doesn't actually live in any layer. *The layers are software layers (Software Agents and Foundation Software Services).* ...bound/glued together by DIDComm-based protocols.

Also read https://github.com/trustoverip/trust-spanning-protocol/discussions/38

**Reply by mwherman2000** [reply]
> I think layer 1 should be labeled _cryptographic_ trust. I don't think DIDComm lives at layer 1.

I agree that Layer 1 Trust Foundation Services is where the basic _cryptographic trust_ services/capabilities live but this isn't the only category of services/capabilities in the Layer 1 Trust Foundation Services layer.

For example, the _DID-CORE CRUDV lifecycle_ is one of the service categories that make up Layer 1 Trust Foundation Services. That is CRUDV of DIDs and DID Documents. The responsible software service is the _DID Registry_.

Similarly, all of the CRUDV (aka Wallety) services/capabilities live as a category the also makes up part of the Layer 1 Trust Foundation Services: the creation (issuance) of VCs, storage of VCs (holding), and verification of VCs, etc.  If you think of the services needed to support the Issuer-Holder-Verifier model, these all live as software services that are part of Layer 1 Trust Foundation Services.  The Proposal 4 Issuer-Holder-Verifier diagram is a Layer 1 diagram: yes, the glue (the protocols) that bind these services and roles together are specializations of the Trust Spanning Layer Base Protocol, they are, in fact, Subprotocols used to access these "lower-level" Layer 1 Trust Foundation Services (as illustrated below).

![image](https://user-images.githubusercontent.com/6101736/227493386-b7915eb1-3395-4061-ad01-61c73c9cc323.png)

Extending this reasoning, support for _virtual web drives_ (e.g. Web5 DWNs, Solid Pods, and Trinity Graphs) all live as a category of _secure storage services_ as part of Layer 1 Trust Foundation Services ...again, communication with these lower-level services is using Subprotocols derived from (specializations of) the Layer 2 Trust Spanning Layer Base Protocol (i.e. built on top of DIDComm Messaging).

Lastly (for this post), everything to do with the _DIDComm Network_ capabilities: mediators, message routing, message relays, support for disconnected agents (behind a firewall or agents that are simply not running or are otherwise unreachable), etc. are all Layer 1 Trust Foundation Services access through specializations of the Trust Spanning Layer Base Protocol (i.e. built on top of DIDComm Messaging).

All of these categories of Layer 1 Trust Foundation Services are illustrated below in this Proposal 4 Appendix A version 0.30 slide:

![image](https://user-images.githubusercontent.com/6101736/227719972-1d0e7524-c5d6-43ae-9dbe-9c170775fce5.png)

All communication is based on derivations (specializations) of the Trust Spanning Layer Base Protocol. Period.  When these communications are accessing "lower-level" Layer 1 Trust Foundation Services (including intra-service layer communications) these are refered to as Layer 1 Trust Foundation Services Subprotocols.

**Comment by mwherman2000** [comment]
To summarize, only 2 layers are needed for a Trust Spanning Layer Protocol model:
- Layer 4 Trust Agents (Software) Layer
- Layer 1 Trust Foundation Services (Software) Layer

...as illustrated in the following _Web 7.0 Two--Layer Honey-Peanut Butter-Jelly (HPBJ) Model for Trust Spanning Layer Frameworks_...

![image](https://user-images.githubusercontent.com/6101736/227775584-b5c73ead-ac57-48a5-aca0-11e5c9a966dd.png)

Reference: Appendix D of Proposal 4 version 0.31 slide 118

The Trust Spanning Layer Base Protocol is assumed to be based on the _Web 7.0 Credential Sender-Receiver Pattern_... 

![image](https://user-images.githubusercontent.com/6101736/227779373-f6f869d8-625a-4cdf-b42b-fd991887aa5e.png)

Reference: Proposal 4 version 0.31 slide 99

**Comment by mwherman2000** [comment]
A significant elaboration and extension of these concepts can be found here: https://github.com/trustoverip/trust-spanning-protocol/discussions/34#discussioncomment-5443632
